### PR TITLE
ObsStore type conversion support

### PR DESCRIPTION
**Pull request by rhoneyager-tomorrow**
## Description

This fixes a longstanding bug in ioda where the ObsStore engine could not convert between different data types. The data type used to store data had to be the same type used to access data.

This was problematic for a variety of reasons:
- Fortran could not read anything written using an unsigned integral type.
- This blocked full implementation of the JEDI Data Conventions.
- Any changes in a variable's type, such as from a float to an int, required a lot of debugging to find every place where that variable was used.

The new feature uses HDF5's H5Tconvert function to perform the conversion. Conversion only happens where needed, so this has a marginal impact on ioda's overall performance.

**Note:** This PR builds on JCSDA/ioda#14, so review that one first. GitHub's UI does a poor job of linking base and head repositories, and I cannot make a PR in JCSDA/ioda that depends on a branch in my own fork. 

## Dependencies

List the other PRs that this PR is dependent on:
- [x] JCSDA/ioda#14

## Impact

Expected impact on downstream repositories: happier downstream developers.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have run the unit tests before creating the PR

### CODE REVIEW

**Comment by srherbener** [comment]
@rhoneyager-tomorrow are you saying in your initial description that PR #14 needs to be merged before this one? Thanks

File: src/engines/ioda/src/ioda/Engines/HH/HH-util.cpp
Comment (srherbener) [inline src/engines/ioda/src/ioda/Engines/HH/HH-util.cpp]: Please remove commented code, unless it will be useful to keep this in. Thanks

File: src/engines/ioda/src/ioda/Engines/HH/HH-util.cpp
Comment (srherbener) [inline src/engines/ioda/src/ioda/Engines/HH/HH-util.cpp]: Please remove commented code, unless it's useful to keep it in.

**Comment by rhoneyager-tomorrow** [comment]
> @rhoneyager-tomorrow are you saying in your initial description that PR #14 needs to be merged before this one? Thanks

Yes. This PR does use the `ioda::detail::Engines::HH::hdf5_error_check()` function introduced in #14.

File: src/engines/ioda/src/ioda/Engines/ObsStore/Type.cpp
Comment (srherbener) [inline src/engines/ioda/src/ioda/Engines/ObsStore/Type.cpp:116]: I'm guessing this block of commented code is a placeholder for a future upgrade. If so, I'm good keeping it in.

**Review by srherbener — COMMENTED** [review]
Thanks for doing this upgrade!

Did this feature branch get the feature branch from #14 merged in? If so, sorry about the repeated comments.

These changes create a dependency between the HH and ObsStore backends. I'm not crazy about this since to me it seems that the HH and ObsStore backends should be independent of each other. However, I see the convenience of using the hdf5 conversion  utilities and I think the hdf5 library is here to stay. How feasible and how much effort do you think it would be to create a conversion utility out of just hdf5 library calls, where that utility could be shared between HH and ObsStore. What are your thoughts about this?

Thanks!

File: src/engines/ioda/src/ioda/Engines/HH/HH-util.cpp
Comment (rhoneyager-tomorrow) [reply src/engines/ioda/src/ioda/Engines/HH/HH-util.cpp]: Ah. This is in #14, which I haven't merged in yet.

File: src/engines/ioda/src/ioda/Engines/ObsStore/Type.cpp
Comment (rhoneyager-tomorrow) [reply src/engines/ioda/src/ioda/Engines/ObsStore/Type.cpp:116]: Yes, it's a placeholder for any further changes to the type system.

Comment (srherbener) [reply src/engines/ioda/src/ioda/Engines/ObsStore/Type.cpp:116]: Thanks for the explanation. I think it's okay to keep this in, no need to change this.

File: src/engines/ioda/src/ioda/Engines/HH/HH-util.cpp
Comment (srherbener) [reply src/engines/ioda/src/ioda/Engines/HH/HH-util.cpp]: This was addressed in #14 already, so just need to make sure the merging stays with the version in #14. Thanks.

File: src/engines/ioda/src/ioda/Engines/HH/HH-util.cpp
Comment (srherbener) [reply src/engines/ioda/src/ioda/Engines/HH/HH-util.cpp]: No need to change since it has been addressed in #14. Thanks.

**Comment by rhoneyager-tomorrow** [comment]
> These changes create a dependency between the HH and ObsStore backends. I'm not crazy about this since to me it seems that the HH and ObsStore backends should be independent of each other. However, I see the convenience of using the hdf5 conversion utilities and I think the hdf5 library is here to stay. How feasible and how much effort do you think it would be to create a conversion utility out of just hdf5 library calls, where that utility could be shared between HH and ObsStore. What are your thoughts about this?
> 
> Thanks!

These type conversions are implicit in the hdf5 library calls, as `H5Tconvert` is internally used within `H5Aread`, `H5Awrite`, `H5Dread`, and `H5Dwrite`. So, the conversion facility would only be used by ObsStore.

The main crossover between HH and ObsStore is the type object conversion from the ObsStore-defined type and the HDF5-equivalent. All of the convenience functions (e.g. for correct HDF5 object construction/destruction, for creation of string types) exist in HH, and I don't think that it makes much sense to reimplement in a more generic manner.

**Comment by srherbener** [comment]
@rhoneyager-tomorrow thanks for the explanation. I see the issues with trying to create a generic utility and agree that such a utility would only be useful for ObsStore.

The tradeoff of using an existing, good conversion utility (H5Tconvert) versus implementing a similar utility for ObsStore is what we are looking at, and I think in a pragmatic sense that using the H5Tconvert utility is a better choice. I also recall that ObsStore was meant to be a replacement for the hdf5 memory file backend which has some performance issues, and discussions about how ObsStore could have been more of a layer built on hdf5 instead of a completely independent backend. In that context I think the notion of a fully independent backend for ObsStore is not as critical as making the other backends (ODB, BUFR) independent.

**Review by srherbener — APPROVED** [review]

**Comment by rhoneyager-tomorrow** [comment]
> In that context I think the notion of a fully independent backend for ObsStore is not as critical as making the other backends (ODB, BUFR) independent.

It is challenging to implement a generic type system well, and that's why HDF5's type system is a great feature to preserve. ObsStore was designed only for a local system's in-memory use, and that made things a bit simpler. If you want to go beyond this, then I suggest using HDF5's type system implementation everywhere instead of making another type system backend. There are just so many pitfalls:

- Cross-platform type support - Equivalent fundamental types do not always exist across system architectures.
    - For integers, how are negative values represented at the bit level? Two's complement is common, but one's complement or signed magnitude approaches are possible. 
    - For floats, processors do not strictly follow [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754). I actually ran into this when importing data from an older NOAA machine. Floating point values may contain different widths of bitfields to represent mantissas and exponents. There are multiple ways to express 16-bit floats: [bfloat16](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format) and [binary16](https://en.wikipedia.org/wiki/Half-precision_floating-point_format).
    - Endianness - AMD64 machines are little-endian. ARM is bi-endian. Many other processor families are big-endian.
- Cross-language compatibility - Equivalent fundamental and compound types do not always exist in C++, C, Fortran, and Python. Exs: unsigned values do not exist in Fortran. Python has built-in support for half-precision floats. How are strings implemented in each language? Ex: Fortran defaults to using fixed-width, space-padded strings. Many, but not all standard C library functions assume null-termination. C++'s `std::string` uses a size + variable-length array (null padding optional).

**Comment by climbfuji** [comment]
@rhoneyager-tomorrow this PR now has conflicts, can you update please?

@ctgh Do you have time to review this? Thanks very much!

**Comment by ctgh** [comment]
@climbfuji I am happy to review this although I would prefer to wait until the conflicts have been resolved.

**Comment by rhoneyager-tomorrow** [comment]
Resolved. The conflicts all were around the removal of a few lines of commented out code.

File: src/engines/Examples/Basic/03-VariablesIntro.cpp
Comment (ctgh) [inline src/engines/Examples/Basic/03-VariablesIntro.cpp]: You could remove this.

File: src/engines/ioda/src/ioda/Engines/ObsStore/Attributes.cpp
Comment (ctgh) [inline src/engines/ioda/src/ioda/Engines/ObsStore/Attributes.cpp:64]: I believe these can be `const`, as can the following `size_t` variables.

File: src/engines/ioda/src/ioda/Engines/ObsStore/Attributes.cpp
Comment (ctgh) [inline src/engines/ioda/src/ioda/Engines/ObsStore/Attributes.cpp:59]: This code block is repeated (with some variation) several times. Is it worth using a common utility function to reduce duplication?

**Review by ctgh — COMMENTED** [review]
Thanks for adding this feature! I have a few minor comments.

File: src/engines/Examples/Basic/03-VariablesIntro.cpp
Comment (rhoneyager-tomorrow) [inline src/engines/Examples/Basic/03-VariablesIntro.cpp]:
```suggestion
```

File: src/engines/ioda/src/ioda/Engines/ObsStore/Attributes.cpp
Comment (rhoneyager-tomorrow) [reply src/engines/ioda/src/ioda/Engines/ObsStore/Attributes.cpp:59]: It's varied four times, and there shouldn't be any more cases. Twice for writes, twice for reads. Attributes and datasets have slightly different semantics, as attributes do not support partial writes/reads. Reads vs writes have similar operations, though they in a different order (write: copy data to temp buffer, convert, then pass along to the internal storage object; read: get from internal storage, convert, then copy to client buffer). Each case has 12 lines of functional code, and I'm not sure if there is much savings overall.

**Comment by climbfuji** [comment]
Quick follow up on this PR, have all review comments been addressed?

**Review by ctgh — APPROVED** [review]
Setting the variables `const` would be nice to have but I don't want it to hold up the merge.

**Comment by climbfuji** [comment]
> Setting the variables `const` would be nice to have but I don't want it to hold up the merge.

@rhoneyager-tomorrow Can you make this small change before I take your PR for merging to JCSDA-internal, please?

**Comment by climbfuji** [comment]
> > Setting the variables `const` would be nice to have but I don't want it to hold up the merge.
> 
> @rhoneyager-tomorrow Can you make this small change before I take your PR for merging to JCSDA-internal, please?

Don't worry about the small `const` change @rhoneyager-tomorrow. I'll add it when I bring your PR over to JCSDA-internal, and have Steve and Chris review it.

File: src/engines/ioda/src/ioda/Engines/ObsStore/Attributes.cpp
Comment (climbfuji) [reply src/engines/ioda/src/ioda/Engines/ObsStore/Attributes.cpp:64]: Done in https://github.com/JCSDA-internal/ioda/pull/1228

**Comment by climbfuji** [comment]
This was merged as https://github.com/JCSDA-internal/ioda/pull/1228

**Pull request closed without merging**
